src/zap.ts

```diff
@@ -0,0 +1,55 @@
+import { ChainId, isEthereum, isFantom } from "./chain";
+import { EthAddress, ZeroAddress } from "./helpers";
+import { Token, VaultMetadataOverrides } from "./types";
+
+export type ZapInWith = keyof Token["supported"];
+
+type ZapInDetails = { isZapInSupported: boolean; zapInWith?: ZapInWith };
+
+type ZapInArgs = {
+  chainId: ChainId;
+  token?: Partial<Token>;
+  vaultMetadata: VaultMetadataOverrides | null;
+};
+
+export function getZapInDetails({ chainId, token, vaultMetadata }: ZapInArgs): ZapInDetails {
+  if (!token?.supported || !vaultMetadata?.zapInWith) {
+    return { isZapInSupported: false };
+  }
+
+  if (isEthereum(chainId)) {
+    if (token.address === EthAddress) {
+      return { isZapInSupported: Boolean(token.supported.zapper), zapInWith: "zapperZapIn" };
+    }
+
+    const { zapInWith } = vaultMetadata;
+
+    if (zapInWith && isValidZap(zapInWith, token.supported)) {
+      const isZapInSupported = Boolean(token.supported.zapper) && Boolean(token.supported[zapInWith]);
+      return { isZapInSupported, zapInWith };
+    }
+  }
+
+  if (isFantom(chainId)) {
+    const zapInWith = "ftmApeZap"; // Hardcoded for now
+
+    if (token.address === ZeroAddress) {
+      return { isZapInSupported: Boolean(token.supported[zapInWith]), zapInWith };
+    }
+
+    const isZapInSupported = Boolean(token.supported[zapInWith]);
+    return { isZapInSupported, zapInWith };
```

Comment:
should we return `zapInWith` as `null` when `isZapInSupported === false` ?

Reply:
I thought about this, and I'm not sure. The reality is that the vault metadata supports `zapInWith`, it's just the token that is not zappable in. I'll let you make the call, would that be beneficial in some way?

Reply:
I think of this function as if it returns whether the zap tx is allowed given the token and vault, and through what contract. Since we want to be consistent with the result, then when a zap is not supported, we should also not return a valid contract.

Reply:
Makes sense, refactored to return `null` with https://github.com/yearn/yearn-sdk/pull/280/commits/d50df443892bb95af08dae8523521b9546259d8d